Source/Concurrency/InductiveSequentialization.cs

```diff
@@ -435,6 +440,293 @@ private List<Declaration> GenerateStepChecker(Action pendingAsync)
+      }
+      
+      var cmds = new List<Cmd>();
+      var successM=  Substituter.Apply(substG0,Substituter.Apply(substM, Expr.And(gateM, GetNonBlockExpression(targetAction))));
+      cmds.Add(CmdHelper.AssumeCmd(successM));
+
+      var disjointnessExprTotal = Expr.And(disjointnessExpr.Values);
+      cmds.Add(CmdHelper.AssumeCmd(disjointnessExprTotal));
+
+      var subsetExprTotal = Expr.And(subsetExpr.Values);
+      cmds.Add(CmdHelper.AssumeCmd(subsetExprTotal));
+     
+      cmds.Add(CmdHelper.AssertCmd(
+        targetAction.tok,
+        Expr.Not(Expr.And(Substituter.Apply(substE1, gateE1), Substituter.Apply(substE2, Expr.And(GetExitCondition(E2), gateE2)))) ,
+        $"Inconsistency check failed for {targetAction.Name}, {E1.Name}, {E2.Name}"));
+      
+      CivlUtil.ResolveAndTypecheck(civlTypeChecker.Options, cmds);
+      return GetCheckerTuple($"{rule}_InconsistencyChecker_{targetAction.Name}_{E1.Name}_{E2.Name}", new List<Requires>(),
+        new List<Variable>(), new List<Variable>(), locals,  cmds);
+    }
+    
+    private Expr GetNonBlockExpression(Action action)
+    {
+      HashSet<Variable> frame = new HashSet<Variable>();
+      frame.UnionWith(action.UsedGlobalVarsInGate);
+      frame.UnionWith(action.UsedGlobalVarsInAction);
+      return TransitionRelationComputation.Cooperation(civlTypeChecker, action, frame);
+    }
+
+    private Expr GetExitCondition(Action action)
+    {
+      var subst = Substituter.SubstitutionFromDictionary(
+        action.ActionDecl.InParams.Zip(action.Impl.InParams.Select(x => (Expr)Expr.Ident(x))).ToDictionary(x => x.Item1, x => x.Item2));
+      var exitCondition = action.ExitCondition;
+      return exitCondition == null ? Expr.False : Substituter.Apply(subst, exitCondition);
+    }
+
+    protected List<Declaration> GenerateExitConditionProperty1Checker(Action action)
+    {
+      var eliminatedActionDecls = EliminatedActionDecls.ToHashSet();
+      var elimExprs = new List<Expr>();
+      var notElimExprs = new List<Expr>();
+      foreach (var pendingAsync in action.PendingAsyncs)
+      {
+        var pendingAsyncExpr = Expr.Ident(action.PAs(pendingAsync.PendingAsyncType));
+        var emptyExpr = ExprHelper.FunctionCall(pendingAsync.PendingAsyncConst, Expr.Literal(0));
+        if (eliminatedActionDecls.Contains(pendingAsync))
+        {
+          elimExprs.Add(Expr.Eq(pendingAsyncExpr, emptyExpr));
+        }
+        else
+        {
+          notElimExprs.Add(Expr.Eq(pendingAsyncExpr, emptyExpr));
+        }
+      }
+      var cmds = new List<Cmd>() {
+        CmdHelper.AssumeCmd(GetExitCondition(action)),
+        CmdHelper.CallCmd(action.Impl.Proc, action.Impl.InParams, action.Impl.OutParams),
+        GetCheck(action.tok, Expr.And(elimExprs), "Exit condition property 1 failed"),
+      };
+
+      //CivlUtil.ResolveAndTypecheck(civlTypeChecker.Options, cmds);
```

Comment:
I tried to uncomment this code but I got a crash. Needs to be investigated.

Reply:
The CallCmd is the problem. It looks like none of the other procedures that have this cmd resolve and type check it. I am not sure why though.

Reply:
I put a fix in the ```ResolveAndTypecheck``` method. The type checking context ```tc``` has a field ```Proc``` which is the reference to the enclosing procedure. This is null and being accessed to do the modifies checking. I disabled the modifies checking trusting that ```GetCheckerTuple``` will set up the modified clause of the enclosing  checker procedure appropriately.